common/datavalues/src/types/serializations/array.rs

```diff
@@ -12,86 +12,91 @@
+        buf.push(b'[');
+        let inner = &self.inner;
+        for i in start..end {
+            if i != start {
+                buf.extend_from_slice(b", ");
             }
-            res.push(']');
-            Ok(res)
-        } else {
-            Err(ErrorCode::BadBytes("Incorrect Array value"))
+            inner.write_field_quoted(i, buf, format, b'\'');
         }
+        buf.push(b']');
     }
 
-    fn serialize_column(&self, column: &ColumnRef, format: &FormatSettings) -> Result<Vec<String>> {
-        let column: &ArrayColumn = Series::check_get(column)?;
-        let mut result = Vec::with_capacity(column.len());
-        for i in 0..column.len() {
-            let val = column.get(i);
-            let s = self.serialize_value(&val, format)?;
-            result.push(s);
+    fn serialize_clickhouse_const(
+        &self,
+        format: &FormatSettings,
+        size: usize,
+    ) -> Result<opensrv_clickhouse::types::column::ArcColumnData> {
+        let len = self.offsets.len() - 1;
+        let mut offsets = opensrv_clickhouse::types::column::List::with_capacity(size * len);
+        let total = self.offsets[len];
+        let mut base = 0;
+        for _ in 0..size {
+            for offset in self.offsets.iter().skip(1) {
+                offsets.push(((*offset) + base) as u64);
+            }
+            base += total;
         }
-        Ok(result)
-    }
 
-    fn serialize_json(&self, column: &ColumnRef, _format: &FormatSettings) -> Result<Vec<Value>> {
-        let column: &ArrayColumn = Series::check_get(column)?;
-        let mut result = Vec::with_capacity(column.len());
-        for i in 0..column.len() {
-            let val = column.get(i);
-            let s = serde_json::to_value(val)?;
-            result.push(s);
-        }
-        Ok(result)
+        let inner_data = self.inner.serialize_clickhouse_const(format, size)?;
+        Ok(Arc::new(ArrayColumnData::create(inner_data, offsets)))
     }
 
-    fn serialize_clickhouse_format(
+    fn serialize_clickhouse_column(
         &self,
-        column: &ColumnRef,
         format: &FormatSettings,
     ) -> Result<opensrv_clickhouse::types::column::ArcColumnData> {
-        let column: &ArrayColumn = Series::check_get(column)?;
-        let mut offsets = opensrv_clickhouse::types::column::List::with_capacity(column.len());
-        for offset in column.offsets().iter().skip(1) {
+        let mut offsets =
+            opensrv_clickhouse::types::column::List::with_capacity(self.offsets.len() - 1);
+        for offset in self.offsets.iter().skip(1) {
             offsets.push(*offset as u64);
         }
 
-        let inner_data = self
-            .inner
-            .serialize_clickhouse_format(column.values(), format)?;
+        let inner_data = self.inner.serialize_clickhouse_column(format)?;
         Ok(Arc::new(ArrayColumnData::create(inner_data, offsets)))
     }
+
+    fn serialize_json(&self, format: &FormatSettings) -> Result<Vec<Value>> {
```

Comment:
So row based api not works on `serialize_json` ?

Reply:
serializer can has API for both row based and col-based(clickhouse,  json::Value)

Reply:
format choose to use which

Reply:
```
pub fn block_to_json_value(
    block: &DataBlock,
    format: &FormatSettings,
) -> Result<Vec<Vec<JsonValue>>> {
    let cols = block_to_json_value_columns(block, format)?;
    Ok(transpose(cols))
}
```

Since we need to transpose, so `serialize_json ` is row based ?

Reply:
But it's not zero alloc in the transpose function. So row based is faster in this case.

Reply:
> But it's not zero alloc in the transpose function. So row based is faster in this case.

make sense

Reply:
may I refactor it later in another pr later,  this pr is large and need to change new-added tests case, hope it to merge early.

Reply:
also need add formats based on this